src/sqlfluff/rules/L003.py

```diff
@@ -315,40 +429,19 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
           indent meta segment in the previous line.
 
         """
-        # Config type hints
-        self.tab_space_size: int
-        self.indent_unit: str
+        # # Config type hints
+        # self.tab_space_size: int
+        # self.indent_unit: str
 
         raw_stack: Tuple[BaseSegment, ...] = context.raw_stack
-
         # We ignore certain types (e.g. non-SQL scripts in functions)
         # so check if on ignore list
-        if context.segment.type in self._ignore_types:
-            return LintResult()
-        for parent in context.parent_stack:
-            if parent.type in self._ignore_types:
+        for element in context.parent_stack + (context.segment,):
+            if element.is_type(*self._ignore_types):
                 return LintResult()
 
         # Memory keeps track of what we've seen
-        if not context.memory:
-            memory: dict = {
-                # in_indent keeps track of whether we're in an indent right now
-                "in_indent": True,
-                # problem_lines keeps track of lines with problems so that we
-                # don't compare to them.
-                "problem_lines": [],
-                # hanging_lines keeps track of hanging lines so that we don't
-                # compare to them when assessing indent.
-                "hanging_lines": [],
-                # comment_lines keeps track of lines which are all comment.
-                "comment_lines": [],
-                # segments we've seen the last child of
-                "finished": set(),
-                # First non-whitespace node on a line.
-                "trigger": None,
-            }
-        else:
-            memory = context.memory
+        memory: _MemoryDict = context.memory or _MemoryDict()
```

Comment:
I like this. I wonder if we should add core support for "memory as a class". The current API encourages you to use dictionaries, which are awkward to build and work with.

Reply:
Perhaps mainly this is for the typing. Which I think is more important than the class. In fact wouldn't be a fan of memory that did anything (ie not a class) just remembered (data)

Reply:
I guess my main point is, I like having the default values for new memory in a canonical, easy to find place, rather than buried somewhere deep inside `_eval()`.

Reply:
Ahhh I've a nice idea for that, why not have a class property of the init value of memory.
then the _crawl could also implement it for you